### PR TITLE
manual_let_else: support struct patterns

### LINKED ISSUE #10424: [manual_let_else] does not take binding variable name into account
### Summary

`manual_let_else` suggestion ignores variable name whether binding pattern have same name or not.

### Reproducer

I tried this code:

```rust
let name = if let Token::Identifier { inner } = ident_token.data {
    inner
} else {
    return Err(SimpleErrorWithPos {
        position: ident_token.position,
        kind: ParserError::UnexpectedToken {
            pat: TokenKind::Identifier,
            unmatch: ident_token.data,
        }
    })
}
```

I expected to see this happen:

```
help: consider writing
    |
498 ~         let Token::Identifier { inner: name } = ident_token.data else {
499 +             return Err(SimpleErrorWithPos {
500 +                 position: ident_token.position,
501 +                 kind: ParserError::UnexpectedToken {
502 +                     pat: TokenKind::Identifier,
503 +                     unmatch: ident_token.data,
504 +                 }
505 +             })
506 +         };
    |
```

Instead, this happened:

```
help: consider writing
    |
498 ~         let Token::Identifier { inner } = ident_token.data else {
499 +             return Err(SimpleErrorWithPos {
500 +                 position: ident_token.position,
501 +                 kind: ParserError::UnexpectedToken {
502 +                     pat: TokenKind::Identifier,
503 +                     unmatch: ident_token.data,
504 +                 }
505 +             })
506 +         };
    |
```

### Version

```text
rustc 1.69.0-nightly (d962ea578 2023-02-26)
binary: rustc
commit-hash: d962ea57899d64dc8a57040142c6b498a57c8064
commit-date: 2023-02-26
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

### Additional Labels

@rustbot label I-suggestion-causes-error

### LINKED ISSUE #10708: manual_let_else incorrectly handling destructuring or-patterns
### Summary

Similar to #9938, clippy now correctly suggests the brackets, but fails to take into account the pattern it's replacing is binding names for further use. 

### Reproducer

I tried this code:

```rust
let data_name = match &data_item {
        Item::Struct(ItemStruct { ident , ..})
        | Item::Enum(ItemEnum { ident , ..})
        | Item::Union(ItemUnion { ident , ..}) => ident,
        _ => unreachable!(),
    };
```

I expected to see this suggested:
```rust
    let (Item::Struct(ItemStruct { ident: data_name, .. })
        | Item::Enum(ItemEnum { ident: data_name, .. })
        | Item::Union(ItemUnion { ident: data_name, .. })) = &data_item else { unreachable!() };
```

Instead, this happened:
```rust
...help: consider writing
   |
69 ~     let (Item::Struct(ItemStruct { ident, .. })
70 +         | Item::Enum(ItemEnum { ident, .. })
71 +         | Item::Union(ItemUnion { ident, .. })) = &data_item else { unreachable!() };
   |
```

### Version

```text
rustc 1.71.0-nightly (7f94b314c 2023-04-23)
binary: rustc
commit-hash: 7f94b314cead7059a71a265a8b64905ef2511796
commit-date: 2023-04-23
host: x86_64-pc-windows-msvc
release: 1.71.0-nightly
LLVM version: 16.0.2
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by est31**
This adds upon the improvements of #10797 and:

* Only prints `()` around `Or` patterns at the top level (fixing a regression of #10797)
* Supports multi-binding patterns: `let (u, v) = if let (Some(u_i), Ok(v_i)) = ex { (u_i, v_i) } else ...`
* Traverses through tuple patterns: `let v = if let (Some(v), None) = ex { v } else ...`
* Supports struct patterns: `let v = if let S { v, w, } = ex { (v, w) } else ...`
 
```
changelog: [`manual_let_else`]: improve pattern printing to support struct patterns
```

fixes #10708
fixes #10424

### CODE REVIEW

**Comment by Manishearth** [comment]
@Centri3 this is an interesting PR; would you like to review it as well? No worries if not.

(trying to point contributors at interesting PRs to review to get more folks who know how to review things)

File: clippy_lints/src/manual_let_else.rs [hunk 1]
```diff
@@ -358,37 +396,59 @@ fn pat_allowed_for_else(cx: &LateContext<'_>, pat: &'_ Pat<'_>, check_types: boo
...
+    let (sub_pats, paths) = match (local_pat.kind, peeled.kind) {
+        (PatKind::Tuple(pats, _), ExprKind::Tup(exprs)) => (pats, exprs),
+        (PatKind::Slice(pats, ..), ExprKind::Array(exprs)) => (pats, exprs),
+        (_, ExprKind::Path(_)) => (slice::from_ref(local_pat), slice::from_ref(peeled)),
+        _ => return None,
     };
+
+    // There is some length mismatch, which indicates usage of .. in the patterns above e.g.:
+    // let (a, ..) = if let [a, b, _c] = ex { (a, b) } else { ... };
+    // For now, bail in these cases.
```
Comment (Manishearth) [inline clippy_lints/src/manual_let_else.rs:449]: thought: seems reasonable; i can't think of any big reason to do this beyond WIP code.

File: clippy_lints/src/manual_let_else.rs [hunk 2]
```diff
@@ -358,37 +396,59 @@ fn pat_allowed_for_else(cx: &LateContext<'_>, pat: &'_ Pat<'_>, check_types: boo
...
+
     let mut pat_bindings = FxHashSet::default();
-    pat.each_binding_or_first(&mut |_ann, _hir_id, _sp, ident| {
+    let_pat.each_binding_or_first(&mut |_ann, _hir_id, _sp, ident| {
         pat_bindings.insert(ident);
     });
     if pat_bindings.len() < paths.len() {
-        return false;
+        // This rebinds some bindings from the outer scope, or it repeats some copy-able bindings multiple
+        // times. We don't support these cases so we bail here. E.g.:
```
Comment (Manishearth) [inline clippy_lints/src/manual_let_else.rs:460]: thought: probably possible to split into two `let`s, probably not worth it.

(Might be interesting to design a different lint that detects when `let (a, b) = ... ` is splittable)

File: clippy_lints/src/manual_let_else.rs [hunk 3]
```diff
@@ -358,37 +396,59 @@ fn pat_allowed_for_else(cx: &LateContext<'_>, pat: &'_ Pat<'_>, check_types: boo
     !has_disallowed
 }
 
-/// Checks if the passed block is a simple identity referring to bindings created by the pattern
-fn expr_is_simple_identity(pat: &'_ Pat<'_>, expr: &'_ Expr<'_>) -> bool {
+/// Checks if the passed block is a simple identity referring to bindings created by the pattern,
```
Comment (Manishearth) [inline clippy_lints/src/manual_let_else.rs:427]: nit: might be worth including an example here

**Review by Manishearth — COMMENTED** [review]
Mostly looks good. I want to have a closer look later, and perhaps wait for @Centri3's review if she is interested.

File: clippy_lints/src/manual_let_else.rs (hunk 3 above)
Comment (est31) [reply clippy_lints/src/manual_let_else.rs:427]: There is an example just below in the body of the function, should I move it?

Comment (Manishearth) [reply clippy_lints/src/manual_let_else.rs:427]: Yep!

**Comment by Centri3** [comment]
Looks good overall, I have nothing to add past what was commented on already

**Pull request merged**